### ISSUE TITLE
Invalid constructor matcher order result in invalid instruction execution.

### ISSUE THREAD

**Issue by rbran** [icicle-emu/icicle-emu#22]
Instruction ARM6_be `vldmia r12!,{d8,d9,d10,d11,d12,d13,d14,d15}` (bytecode `\xEC\xBC\x8B\x10` ) is being parsed/executed as `ldc p11,cr8,[r12],#0x40"`.

That happen because the bytecode is being matched against `ldc` before `vldmia`, although `vldmia` comes first on the slaspec file and `ldc` don't contains the `vldmia` pattern bits.

I was not able to understand why that happen yet, I'll investigate later. First i need to learn how this graphical ordering process of `sort_overlaps` works.

Here a comparison of Sleigh-rs and Icicle-emu. Note that icicle branch the pattern for each `|` in the patter, sleigh-rs unify them.

Constraint in sleigh-rs in order they are checked:
|inst | location | context | token
|---|----|-----|---|
|vldmia | ARMneon:3443 | `XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX1` | `0011XXXXXXXX1XX11101XXXX0XXXXXXX`
|vldmia |ARMneon:3700 | `XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX1` | `0011XXXXXXXX1XX10101XXXXXXXXXXXX`
|ldc | ARMInstructions:2709| `0XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX11` | `X011XXXXXXXX1X0XXXXXXXXXXXXXXXXX`
|ldc | ARMTHUMB:1904 | `1XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX1` | `X0110111XXXX1X0XXXXXXXXXXXXXXXXX`

Constraint in icicle  in order they are checked.
|inst | location | context | token
|------------|------------|-----------|-------
|vldmia  | ARMneon:3700 | `0XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX11` | `0011XXXXXXXX1XX10101XXXXXXXXXXXX`
|ldc  | ARMInstructions:2709|  `0XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX11` | `X011XXXXXXXX1X0XXXXXXXXXXXXXXXXX`
|vldmia  | ARMneon:3700 | `1XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX1` | `00110111XXXX1XX10101XXXXXXXXXXXX`
|vldmia |ARMneon:3443 | `0XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX1` | `0011XXXXXXXX1XX11101XXXX0XXXXXXX`
|vldmia  |ARMneon:3443 | `1XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX1` | `00110111XXXX1XX11101XXXX0XXXXXXX`
|ldc  |ARMTHUMB:1904 | `1XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX1` | `X0110111XXXX1X0XXXXXXXXXXXXXXXXX`

**Comment by rbran** [comment]
Turns out, it had nothing to do with petgraph, if was just a mistake in the match statement that unified multiple ordering results.

Fixed with: https://github.com/icicle-emu/icicle-emu/pull/23

**Comment by mchesser** [comment]
There was a reason for the logic before (although I agree it is not likely not correct), the reason the logic was written was to handle a strange case I observed in the x64 specification. The problematic lines occur here: https://github.com/NationalSecurityAgency/ghidra/blob/9ec89c37fb30d6228ed82d4cf1441b7a7aa0a3f1/Ghidra/Processors/x86/data/languages/ia.sinc#L743C1-L746

Which correspond to the following:

```
(743)addr64: [Base64]      is mod=0 & r_m=4; rexXprefix=0 & index64=4    & Base64
(745)addr64: [Index64*ss]  is mod=0 & r_m=4;                Index64      & base64=5 & ss ; imm32=0
(756)addr64: [simm32_64]   is mod=0 & r_m=4; rexXprefix=0 & index64=4    & base64=5      ; simm32_64
```

With the corrected ordering results we get

* 756 should be matched before 743: has strictly has more bits set.
* 743 should be matched before 745: different bits are constrained and it is ordered first.
* 745 should be matched before 756: different bits are constrained and it is ordered first.

This leads to an ordering cycle:

756 ---> 743 ---> 745 ---> 756.


I'm not sure what I'm missing, that allows this cycle to be resolved, so for now the approach I used in 75f0259 was to use a heuristic that prefers that are in a cycle because they have more constraints, over constructors that are in the cycle because of ordering reasons. This seemed to work for all the ordering issues I know about -- but there could be some more cases I am missing.

**Comment by rbran** [comment]
I'm pretty sure the cycle is caused by this [`None`](https://github.com/icicle-emu/icicle-emu/blob/75f0259dfb264e3ffd7d2ecd1f9ede3a9a65d9d9/sleigh/sleigh-compile/src/matcher/mod.rs#L270) using the file declaration order to affect the graph.

I think is possible to revert https://github.com/icicle-emu/icicle-emu/commit/75f0259dfb264e3ffd7d2ecd1f9ede3a9a65d9d9 and make this cycle to be unreachable.
